### PR TITLE
Organization onboarding 

### PR DESCRIPTION
**Pull request by colby-swandale**
# What's this about?

In #4859, we added Organizations to rubygems.org to allow teams and businesses to manage their gems and users under a single umbrella. This Pull Request is the next major part of the work to add Organizations, allowing users to migrate their gems to an Organization.

Introducing Organization Onboarding!

## Details

This Pull Request adds a new model called Organization Onboarding that collects organization field, like name and handle, gems (which the user must be an owner of), and users+roles (who must have ownership of the selected gems).
When the user has confirmed all information, the Onboarding then performs the following steps:

* Links the selected RubyGems to the Organization
* Creates Membership records that link a User to an Organization
* Creates the Organization
* Removes the previously held Ownership records

## Notables
* One of the selected roles for Users is an Outside Collaborator. This is not a role and is, in fact, an option a user can select to ask for the user to not be added to the Organization.
* We initially introduced the concept of a Team & Team Members, but this has been removed for now and will circle back in the near future.

## Screenshots

_initial screen to set the name of the organization_

Screenshot 2024-11-14 at 2 22 13 AM

_gem selection screen_
Screenshot 2024-11-14 at 2 27 50 AM

_user & role selection screen_
Screenshot 2024-11-14 at 2 28 01 AM

_final confirmation screen_
Screenshot 2024-11-14 at 2 28 14 AM

_

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -11,6 +11,7 @@ gem "bootsnap", "~> 1.18"
 gem "clearance", "~> 2.9"
 gem "dalli", "~> 3.2"
 gem "datadog", "~> 2.4"
+gem "debug"
```
Comment (martinemde) [inline Gemfile:14]: Intend for this to stay? move to dev probably?

Comment (colby-swandale) [reply Gemfile:14]: This is for dev (I'll probably introduce this as well soon)

Comment (colby-swandale) [reply Gemfile]: Removed

File: app/models/organization.rb [hunk 2]
```diff
@@ -6,17 +6,14 @@ class Organization < ApplicationRecord
...
-  validate :unique_with_user_handle
-
-  def unique_with_user_handle
-    errors.add(:handle, "has already been taken") if handle && User.where("lower(handle) = lower(?)", handle).any?
-  end
 
   has_many :memberships, -> { where.not(confirmed_at: nil) }, dependent: :destroy, inverse_of: :organization
   has_many :unconfirmed_memberships, -> { where(confirmed_at: nil) }, class_name: "Membership", dependent: :destroy, inverse_of: :organization
   has_many :users, through: :memberships
+  has_many :teams, dependent: :destroy
```
Comment (martinemde) [inline app/models/organization.rb:13]: teams leftover.

**Review by martinemde — APPROVED** [review]
I looked over everything and I only see a few little things to fix or check.

File: app/models/rubygem.rb [hunk 3]
```diff
@@ -430,4 +432,12 @@ def bulk_reorder_versions
     sanitized_query = ActiveRecord::Base.send(:sanitize_sql_array, update_query)
     ActiveRecord::Base.connection.execute(sanitized_query)
   end
+
+  def owned_by_organization?
+    organization.present?
+  end
+
+  def user_authorized_for_organization?(user)
+    organization.memberships.exists?(user: user)
```
[the remark this answers is not in the document]
Comment (colby-swandale) [reply app/models/rubygem.rb:441]: Agreed, I figured this was easier and simpler for this PR, but let's move these authorization checks to Pundit soon.

**Review by martinemde — APPROVED** [review]
💪

**Pull request merged**